HD random scheme key (de)serialization

rvl opened the pull request:
Relates to issue #557 

# Overview

This implements the `instance PersistKey RndKey`.

Because `RndKey` has extra fields, serialization becomes a little more complicated. To keep things simple I restricted the deserializers to `RootK` only.

I also needed to remove the use of deserialize to generate fake addresses (transaction size estimation), which is probably a good thing.

rvl committed:
HD random scheme key (de)serialization

This implements the instance PersistKey RndKey.

Because RndKey has extra fields, serialization becomes a little more
complicated. To keep things simple I restricted the deserializers to
RootK only.

I also needed to remove the use of deserialize to generate fake
addresses (transaction size estimation), which is probably a good
thing.

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Sequential.hs:
```diff
@@ -347,16 +344,17 @@ deserializeXPrvSeq (k, h) = (,)
 deserializeXPubSeq :: ByteString -> Either String (SeqKey purpose XPub)
 deserializeXPubSeq = fmap SeqKey . (xpub <=< fromHexText)
 
+toHexText :: ByteString -> ByteString
```
we should reuse https://github.com/input-output-hk/cardano-wallet/blob/483abd4b920756fbe6605d0a72e37e3e2eb874a9/lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Common.hs#L37 instead of introducing this one again

paweljakubas replied:
ah, ok, you removed the common file - sorry only now detected this

paweljakubas reviewed (approved):
LGTM!

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs:
```diff
@@ -263,6 +270,64 @@ deriveAddressPrivateKey (Passphrase pwd) accountKey idx@(Index addrIx) = RndKey
+    _ -> Left "Key input must be two hex strings separated by :"
+
+serializeXPrvRnd
+    :: (RndKey depth XPrv, Hash "encryption")
```
Wouldn't it be more correct / type-safe to have `RndKey 'RootK XPrv` here? Instead of leaving the depth parameter free. Since we are discarding the derivation path from the serialized value, it wouldn't make sense to allow serializing a `RndKey 'AddressK XPrv` via this method. Don't you think ?

rvl replied:
Yep, fixed.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs:
```diff
@@ -263,6 +270,64 @@ deriveAddressPrivateKey (Passphrase pwd) accountKey idx@(Index addrIx) = RndKey
+-- | The reverse of 'serializeXPub'. This will fail if the input is not a valid
+-- hexadecimal string of the correct length.
+deserializeXPubRnd :: ByteString -> Either String (RndKey 'RootK XPub)
+deserializeXPubRnd = fmap mkKey . (deserializeKey xpub)
```
Note that I think we only need to serialize public key in order to serialize `AddressPool` in the sequential derivation. I don't think we have to serialize / store any random public keys actually :thinking: ?

rvl replied:
I should have checked where it was used. That is right, public keys only need to be serialized for SeqState. Fixed.

rvl committed:
Remove (de)serializeXPub from WalletKey

It's only needed for SeqState address pools of SeqKey.

And also specialize serializeXPrv to RootK.

rvl committed:
Merge branch 'master' into rvl/557/persist-key-rnd
